geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/process/RouteSearchProcess.scala

```diff
@@ -0,0 +1,349 @@
+      val name = Option(geomField).getOrElse(sft.getGeometryDescriptor.getLocalName)
+      val descriptor = sft.getDescriptor(name)
+      if (descriptor == null) {
+        throw new IllegalArgumentException(s"Geometry field '$name' does not exist in input feature collection")
+      }
+      val binding = descriptor.getType.getBinding
+      val isPoints = classOf[Point].isAssignableFrom(binding)
+      if (!isPoints && !classOf[LineString].isAssignableFrom(binding)) {
+        throw new IllegalArgumentException(s"Geometry field '$name' must be a Point or LineString")
+      }
+      (name, isPoints)
+    }
+
+    if (headingField == null && isPoints) {
+      throw new IllegalArgumentException("Heading must be specified unless input feature collection geometry is a LineString")
+    } else if (headingField != null && sft.indexOf(headingField) == -1) {
+      throw new IllegalArgumentException(s"Heading field '$headingField' does not exist in input feature collection")
+    }
+
+    // extract the route geometries
+    val routeGeoms = {
+      val sft = routes.getSchema
+      val name = Option(routeGeomField).getOrElse(sft.getGeometryDescriptor.getLocalName)
+      val index = sft.indexOf(name)
+      if (index == -1) {
+        throw new IllegalArgumentException(s"Geometry field '$name' does not exist in route feature collection")
+      }
+      if (!classOf[LineString].isAssignableFrom(sft.getDescriptor(index).getType.getBinding)) {
+        throw new IllegalArgumentException(s"Route geometry field '$name' must be a LineString")
+      }
+      SelfClosingIterator(routes).map(_.getAttribute(index).asInstanceOf[LineString]).toSeq
+    }
+
+    val visitor = new RouteVisitor(routeGeoms, bufferSize, headingThreshold, bi, geomAttribute, isPoints, Option(headingField))
+
+    features.accepts(visitor, new NullProgressListener)
+    val results = visitor.getResult.results
+
+    configureOutput(features.getSchema, results, bins, binGeom, binDtg, binTrackId, binLabel)
+
+    results
+  }
+}
+
+class RouteVisitor(routes: Seq[LineString],
+                   routeBuffer: Double,
+                   threshold: Double,
+                   bidirectional: Boolean,
+                   geomAttribute: String,
+                   isPoints: Boolean,
+                   headingAttribute: Option[String]) extends FeatureCalc with FeatureAttributeVisitor with LazyLogging {
+
+  private var resultCalc: RouteResult = _
+
+  private val routeFilter =
+    orFilters(routes.map(ls => ff.dwithin(ff.property(geomAttribute), ff.literal(ls), routeBuffer, "meters")))
+
+  // for manual check, rewrite the filter to handle meters
+  // geotools doesn't handle this, we normally do it in our query planner
+  private lazy val manualRouteFilter = routeFilter match {
+    case f: DWithin => new QueryPlanFilterVisitor(null).visit(f, null).asInstanceOf[Filter]
+    case f: Or      => new QueryPlanFilterVisitor(null).visit(f, null).asInstanceOf[Filter]
```

Comment:
the regular filter gets passed into our query planner during execution, and the dwithin is handled there - so we don't want to modify it first. This filter is used directly when manually iterating through the features, so we have to fix the dwithin ourselves. Unfortunately FilterVisitor doesn't have a generic `visit(Filter)` method, so we have to use the match to call the appropriate filter-type-specific method.